velox/functions/prestosql/tests/ArrayShuffleTest.cpp

```diff
@@ -0,0 +1,316 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <folly/container/F14Set.h>
+#include <algorithm>
+#include <numeric>
+#include <optional>
+#include <random>
+#include <sstream>
+
+#include "velox/expression/VectorReaders.h"
+#include "velox/functions/prestosql/tests/utils/FunctionBaseTest.h"
+
+using namespace facebook::velox;
+using namespace facebook::velox::test;
+using namespace facebook::velox::functions::test;
+
+namespace {
+template <typename T>
+std::string printArray(const std::vector<std::optional<T>>& input) {
+  std::stringstream out;
+  out << "[";
+  std::for_each(input.begin(), input.end(), [&out](const auto& e) {
+    if (e.has_value()) {
+      out << e.value();
+    } else {
+      out << "NULL";
+    }
+    out << ", ";
+  });
+  out << "]";
+  return out.str();
+}
+
+template <typename T>
+std::string printArray(
+    const std::vector<std::optional<std::vector<std::optional<T>>>>& input) {
+  std::stringstream out;
+  out << "[";
+  std::for_each(input.begin(), input.end(), [&out](const auto& e) {
+    if (e.has_value()) {
+      out << printArray(e.value());
+    } else {
+      out << "NULL";
+    }
+    out << ", ";
+  });
+  out << "]";
+  return out.str();
+}
+
+template <typename T>
+std::string printArray(const std::vector<T>& input) {
+  std::stringstream out;
+  out << "[";
+  std::for_each(
+      input.begin(), input.end(), [&out](const auto& e) { out << e << ", "; });
+  out << "]";
+  return out.str();
+}
+
+template <typename T, bool returnsOptionalValues>
+exec::ArrayView<returnsOptionalValues, T> read(
+    exec::VectorReader<Array<T>>& reader,
+    size_t offset) {
+  if constexpr (returnsOptionalValues) {
+    return reader[offset];
+  } else {
+    return reader.readNullFree(offset);
+  }
+}
+} // namespace
+
+namespace {
+class ArrayShuffleTest : public FunctionBaseTest {
+ protected:
+  constexpr static const int32_t kNumShuffleTimes = 100;
+  constexpr static const int32_t kNums = 10;
+  constexpr static const double kUniquenessRate = .7;
+
+  template <typename T, bool returnsOptionalValues = false>
```

Comment:
All the calls to testShuffle specify returnsOptionalValues as true. Hence, this template parameter is not needed. Let's remove it.

Reply:
Makes sense. 
FYI: my original idea is to have only one testShuffle() function instead of two functions: testShuffle() + testShuffleRandomness(). Somehow, I need to deal with build errors caused by 1) having std::optional and nested std::vector in a folly::F14FastSet and 2) calculating the right sizes & thresholds for different shape of input array cases. Then, I ended up choosing to separate into two functions to keep brevity.